### PR TITLE
Add e-shop-website-with-ilb-ase

### PR DESCRIPTION
**Pull request by OGcanviz**
### Best Practice Checklist
Check these items before submitting a PR... See the Contribution Guide for the full detail: https://github.com/Azure/azure-quickstart-templates/blob/master/1-CONTRIBUTION-GUIDE/README.md 

1. uri's compatible with all clouds (Stack, China, Government)
1. Staged artifacts use _artifactsLocation & _artifactsLocationSasToken
1. Use resourceGroup().location for resource locations
1. Folder names for artifacts (nestedtemplates, scripts, DSC)
1. Use literal values for apiVersion (no variables)
1. Parameter files (GEN-UNIQUE for value generation and no "changemeplease" values
1. $schema and other uris use https
1. Use uniqueString() whenever possible to generate names for resources.  While this is not required, it's one of the most common failure points in a deployment. 
1. Update the metadata.json with the current date

For details: https://github.com/Azure/azure-quickstart-templates/blob/master/1-CONTRIBUTION-GUIDE/bp-checklist.md

- [x] - Please check this box once you've submitted the PR if you've read through the Contribution Guide and best practices checklist.

### Changelog

*
*
*

### CODE REVIEW

File: e-shop-website-with-ilb-ase/azuredeploy.json [hunk 1]
```diff
@@ -0,0 +1,888 @@
...
+    "adminAppInternalHostName": "[concat(variables('adminAppName'), '.', parameters('existingASEDNSSuffix'))]",
+    "sourceCodeRepositoryURL": "https://github.com/OGCanviz/e-shop-website-with-ilb-ase"
+  },
+  "resources": [
+    {
+      "apiVersion": "2018-01-01",
+      "type": "Microsoft.Network/virtualNetworks/subnets",
+      "name": "[concat(variables('vnetName'), '/', variables('redisCacheSubnetName'))]",
+      "properties": {
+        "addressPrefix": "192.168.251.0/26"
```
Comment (bmoore-msft) [inline e-shop-website-with-ilb-ase/azuredeploy.json]: is this always valid in every vnet?

File: e-shop-website-with-ilb-ase/azuredeploy.json [hunk 2]
```diff
@@ -0,0 +1,888 @@
...
+      }
+    },
+    {
+      "apiVersion": "2018-01-01",
+      "type": "Microsoft.Network/virtualNetworks/subnets",
+      "name": "[concat(variables('vnetName'), '/', variables('agSubnetName'))]",
+      "properties": {
+        "addressPrefix": "192.168.251.64/26"
```
Comment (bmoore-msft) [inline e-shop-website-with-ilb-ase/azuredeploy.json]: is this always valid in every vnet?

File: e-shop-website-with-ilb-ase/azuredeploy.json [hunk 3]
```diff
@@ -0,0 +1,888 @@
...
+      },
+      "dependsOn": [
+        "[resourceId('Microsoft.Network/virtualNetworks/subnets/', variables('vnetName'), variables('redisCacheSubnetName'))]"
+      ]
+    },
+    {
+      "apiVersion": "2015-05-01",
+      "type": "microsoft.insights/components",
+      "name": "[variables('appInsightsName')]",
+      "location": "East US",
```
Comment (bmoore-msft) [inline e-shop-website-with-ilb-ase/azuredeploy.json]: move this to a variable

File: e-shop-website-with-ilb-ase/azuredeploy.json [hunk 4]
```diff
@@ -0,0 +1,888 @@
...
+      "properties": {
+        "ApplicationId": "[variables('appInsightsName')]",
+        "Request_Source": "IbizaWebAppExtensionCreate"
+      }
+    },
+    {
+      "apiVersion": "2016-03-01",
```
Comment (bmoore-msft) [inline e-shop-website-with-ilb-ase/azuredeploy.json]: update apiVersion

File: e-shop-website-with-ilb-ase/azuredeploy.json [hunk 5]
```diff
@@ -0,0 +1,888 @@
...
+        "reserved": false,
+        "hostingEnvironment": "[parameters('existingAseName')]"
+      },
+      "sku": {
+        "tier": "Isolated",
+        "name": "I1"
+      }
+    },
+    {
+      "apiVersion": "2016-03-01",
```
Comment (bmoore-msft) [inline e-shop-website-with-ilb-ase/azuredeploy.json]: update apiVersion

File: e-shop-website-with-ilb-ase/azuredeploy.json [hunk 6]
```diff
@@ -0,0 +1,888 @@
...
+        "siteConfig": {
+          "alwaysOn": true,
+          "appSettings": [
+            {
+              "name": "ODataServiceBaseUrl",
+              "value": "[concat('http://', variables('apiAppInternalHostName'))]"
+            },
+            {
+              "name": "cdn:url",
+              "value": "[concat('http://', variables('cdnWebAppEndpointName'), '.azureedge.net')]"
```
Comment (bmoore-msft) [inline e-shop-website-with-ilb-ase/azuredeploy.json]: use reference() to get the uri's

File: e-shop-website-with-ilb-ase/azuredeploy.json [hunk 7]
```diff
@@ -0,0 +1,888 @@
...
+          ]
+        },
+        "hostNames": [
+          "[variables('webAppExternalHostName')]",
+          "[variables('webAppInternalHostName')]"
+        ]
+      },
+      "resources": [
+        {
+          "apiVersion": "2015-08-01",
```
Comment (bmoore-msft) [inline e-shop-website-with-ilb-ase/azuredeploy.json]: update apiVersion

File: e-shop-website-with-ilb-ase/azuredeploy.json [hunk 8]
```diff
@@ -0,0 +1,888 @@
...
+      "dependsOn": [
+        "[resourceId('Microsoft.Web/serverfarms/', variables('aspName'))]",
+        "[resourceId('microsoft.insights/components/', variables('appInsightsName'))]",
+        "[resourceId('Microsoft.Storage/storageAccounts', variables('storageAccountName'))]",
+        "[resourceId('Microsoft.Cache/Redis/', variables('redisCacheName'))]",
+        "[resourceId('Microsoft.Web/sites/', variables('apiAppName'))]"
+      ]
+    },
+    {
+      "apiVersion": "2016-03-01",
```
Comment (bmoore-msft) [inline e-shop-website-with-ilb-ase/azuredeploy.json]: update apiVersion

File: e-shop-website-with-ilb-ase/azuredeploy.json [hunk 9]
```diff
@@ -0,0 +1,888 @@
...
+      }
+    },
+    {
+      "apiVersion": "2018-01-01",
+      "type": "Microsoft.Network/virtualNetworks/subnets",
+      "name": "[concat(variables('vnetName'), '/', variables('agSubnetName'))]",
+      "properties": {
+        "addressPrefix": "192.168.251.64/26"
+      },
+      "dependsOn": [
```
Comment (bmoore-msft) [inline e-shop-website-with-ilb-ase/azuredeploy.json]: move all the dependsOn above the properties object

File: e-shop-website-with-ilb-ase/azuredeploy.json [hunk 10]
```diff
@@ -0,0 +1,888 @@
...
+          ]
+        },
+        "hostNames": [
+          "[variables('contentAppExternalHostName')]",
+          "[variables('contentAppInternalHostName')]"
+        ]
+      },
+      "resources": [
+        {
+          "apiVersion": "2015-08-01",
```
Comment (bmoore-msft) [inline e-shop-website-with-ilb-ase/azuredeploy.json]: update apiversion

File: e-shop-website-with-ilb-ase/azuredeploy.json [hunk 11]
```diff
@@ -0,0 +1,888 @@
...
+            "[resourceId('Microsoft.Web/sites/', variables('contentAppName'))]"
+          ]
+        }
+      ],
+      "dependsOn": [
+        "[resourceId('Microsoft.Web/serverfarms/', variables('aspName'))]"
+      ]
+    },
+    {
+      "apiVersion": "2016-03-01",
```
Comment (bmoore-msft) [inline e-shop-website-with-ilb-ase/azuredeploy.json]: update apiVersion - do this throughout (won't comment on the rest)

File: e-shop-website-with-ilb-ase/azuredeploy.json [hunk 12]
```diff
@@ -0,0 +1,888 @@
...
+      "name": "[variables('redisCacheName')]",
+      "location": "[variables('location')]",
+      "properties": {
+        "sku": {
+          "name": "Premium",
+          "family": "P",
+          "capacity": 1
+        },
+        "subnetId": "[resourceId('Microsoft.Network/virtualNetworks/subnets/', variables('vnetName'), variables('redisCacheSubnetName'))]",
+        "staticIP": "192.168.251.62",
```
Comment (bmoore-msft) [inline e-shop-website-with-ilb-ase/azuredeploy.json]: how do we know this is not in use?

File: e-shop-website-with-ilb-ase/azuredeploy.json [hunk 13]
```diff
@@ -0,0 +1,888 @@
...
+          "name": "[variables('cdnStorageEndpointName')]",
+          "location": "[variables('location')]",
+          "properties": {
+            "isHttpAllowed": true,
+            "isHttpsAllowed": true,
+            "origins": [
+              {
+                "name": "Storage",
+                "properties": {
+                  "hostName": "[concat(variables('storageAccountName'), '.blob.core.windows.net')]"
```
Comment (bmoore-msft) [inline e-shop-website-with-ilb-ase/azuredeploy.json]: use reference() to get the uri

File: e-shop-website-with-ilb-ase/azuredeploy.json [hunk 14]
```diff
@@ -0,0 +1,888 @@
...
+              "text/css",
+              "text/javascript",
+              "application/x-javascript",
+              "application/javascript",
+              "application/json",
+              "application/xml"
+            ],
+            "optimizationType": "GeneralWebDelivery",
+            "probePath": null,
+            "originHostHeader": "[concat(variables('storageAccountName'), '.blob.core.windows.net')]"
```
Comment (bmoore-msft) [inline e-shop-website-with-ilb-ase/azuredeploy.json]: use reference() to get the uri

File: e-shop-website-with-ilb-ase/azuredeploy.parameters.json [hunk 15]
```diff
@@ -0,0 +1,30 @@
...
+    "existingASEName": {
+      "value": "GET-PREREQ-existingASEName"
+    },
+    "existingASELocation": {
+      "value": "GET-PREREQ-existingASELocation"
+    },
+    "existingASEDNSSuffix": {
+      "value": "GET-PREREQ-existingASEDNSSuffix"
+    },
+    "dnsSuffix": {
```
Comment (bmoore-msft) [inline e-shop-website-with-ilb-ase/azuredeploy.parameters.json]: remove this and just use a defaultValue in the template

File: e-shop-website-with-ilb-ase/metadata.json [hunk 16]
```diff
@@ -0,0 +1,7 @@
+{
+	"itemDisplayName": "eShop Website with ILB ASE",
+	"description": "An App Service Environment is a Premium service plan option of Azure App Service that provides a fully isolated and dedicated environment for securely running Azure App Service apps at high scale, including Web Apps, Mobile Apps, and API Apps.",
+	"summary": "This is a sample for LoB solution with App Service Environment (ASE).",
+	"githubUsername": "TylerLu",
+	"dateUpdated": "2018-03-09"
```
Comment (bmoore-msft) [inline e-shop-website-with-ilb-ase/metadata.json]: update this

File: e-shop-website-with-ilb-ase/prereqs/prereq.azuredeploy.json [hunk 17]
```diff
@@ -0,0 +1,406 @@
+{
+  "$schema": "http://schema.management.azure.com/schemas/2014-04-01-preview/deploymentTemplate.json#",
+  "contentVersion": "1.0.0.0",
+  "parameters": {
+    "name": {
+      "type": "string",
+      "defaultValue": "<YourName>-eshop-<Date>",
```
Comment (bmoore-msft) [inline e-shop-website-with-ilb-ase/prereqs/prereq.azuredeploy.json]: remove this

File: e-shop-website-with-ilb-ase/prereqs/prereq.azuredeploy.json [hunk 18]
```diff
@@ -0,0 +1,406 @@
...
+  "variables": {
+    "location": "[parameters('location')]",
+    "vnetName": "[concat(parameters('name'), '-vnet')]",
+    "aseSubnetName": "[concat(parameters('name'), '-subnet')]",
+    "subnetRouteTableName": "[concat(parameters('name'), '-route-table')]",
+    "subnetNSGName": "[concat(parameters('name'), '-nsg')]"
+  },
+  "resources": [
+    {
+      "apiVersion": "2015-02-01",
```
Comment (bmoore-msft) [inline e-shop-website-with-ilb-ase/prereqs/prereq.azuredeploy.json]: update apiVersions throughout

File: e-shop-website-with-ilb-ase/.gitignore [hunk 19]
```diff
@@ -0,0 +1,290 @@
+## Ignore Visual Studio temporary files, build results, and
```
[the remark this answers is not in the document]
Comment (TylerLu) [reply e-shop-website-with-ilb-ase/.gitignore]: removed

File: e-shop-website-with-ilb-ase/README.md [hunk 20]
```diff
@@ -0,0 +1,239 @@
+# eShop Website with ILB ASE
+
+**Table of content**
```
[the remark this answers is not in the document]
Comment (TylerLu) [reply e-shop-website-with-ilb-ase/README.md]: fixed

File: e-shop-website-with-ilb-ase/README.md [hunk 21]
```diff
@@ -0,0 +1,239 @@
...
+![](images/architecture.png)
+
+Download the Visio diagram [here](/e-shop-website-with-ilb-ase/architecture.vsdx).
+
+## Prerequisites
+
+To experience this sample, you should prepare two domain names:
+
+1. Internal domain name: used by the App Service Environment. For example `contoso-internal.com`.
+2. External domain name: used to internet access Web Apps. For example `contoso.com`.
```
[the remark this answers is not in the document]
Comment (TylerLu) [reply e-shop-website-with-ilb-ase/README.md]: fixed

File: e-shop-website-with-ilb-ase/README.md [hunk 22]
```diff
@@ -0,0 +1,239 @@
...
+
+   | API                            | Permission Type | Permissions                   |
+   | ------------------------------ | --------------- | ----------------------------- |
+   | Windows Azure Active Directory | Delegated       | Sign in and read user profile |
+
+Copy the **Application Id** and save it in a text file. You will need it later to specify the value for the **Azure Ad Client Id** parameter in the ARM Template.
+
+### Deploy the Other Azure Resources
+
+After the App Service Environment deployed, follow the steps below to deploy the other Azure resources.
```
[the remark this answers is not in the document]
Comment (TylerLu) [reply e-shop-website-with-ilb-ase/README.md]: fixed

File: e-shop-website-with-ilb-ase/azuredeploy.json [hunk 23]
```diff
@@ -0,0 +1,888 @@
+{
+  "$schema": "http://schema.management.azure.com/schemas/2014-04-01-preview/deploymentTemplate.json#",
+  "contentVersion": "1.0.0.0",
+  "parameters": {
+    "existingASEName": {
+      "type": "string",
+      "defaultValue": "<YourName>-eshop-<Date>",
```
[the remark this answers is not in the document]
Comment (TylerLu) [reply e-shop-website-with-ilb-ase/azuredeploy.json]: removed

File: e-shop-website-with-ilb-ase/azuredeploy.json [hunk 24]
```diff
@@ -0,0 +1,888 @@
...
+    },
+    "dnsSuffix": {
+      "type": "string",
+      "metadata": {
+        "description": "Root domain name used to access web apps on internet."
+      }
+    },
+    "sqlServerAdministratorLogin": {
+      "type": "string",
+      "defaultValue": "eshop-admin",
```
[the remark this answers is not in the document]
Comment (TylerLu) [reply e-shop-website-with-ilb-ase/azuredeploy.json]: removed

File: e-shop-website-with-ilb-ase/azuredeploy.json [hunk 25]
```diff
@@ -0,0 +1,888 @@
...
+    "azureAdTenantId": {
+      "type": "string",
+      "defaultValue": "",
+      "metadata": {
+        "description": "Directory ID of the Azure Active Directory used by the Admin App."
+      }
+    },
+    "azureAdClientId": {
+      "type": "string",
+      "defaultValue": "",
```
[the remark this answers is not in the document]
Comment (TylerLu) [reply e-shop-website-with-ilb-ase/azuredeploy.json]: yes, it is required. Null is not a valid value. I removed the defaultValue.

File: e-shop-website-with-ilb-ase/azuredeploy.json (hunk 3 above)
Comment (TylerLu) [reply e-shop-website-with-ilb-ase/azuredeploy.json]: done

File: e-shop-website-with-ilb-ase/azuredeploy.json (hunk 4 above)
Comment (TylerLu) [reply e-shop-website-with-ilb-ase/azuredeploy.json]: updated

File: e-shop-website-with-ilb-ase/azuredeploy.json (hunk 5 above)
Comment (TylerLu) [reply e-shop-website-with-ilb-ase/azuredeploy.json]: updated

File: e-shop-website-with-ilb-ase/azuredeploy.json (hunk 7 above)
Comment (TylerLu) [reply e-shop-website-with-ilb-ase/azuredeploy.json]: updated

File: e-shop-website-with-ilb-ase/azuredeploy.json (hunk 8 above)
Comment (TylerLu) [reply e-shop-website-with-ilb-ase/azuredeploy.json]: updated

File: e-shop-website-with-ilb-ase/azuredeploy.json (hunk 9 above)
Comment (TylerLu) [reply e-shop-website-with-ilb-ase/azuredeploy.json]: done

File: e-shop-website-with-ilb-ase/azuredeploy.json (hunk 10 above)
Comment (TylerLu) [reply e-shop-website-with-ilb-ase/azuredeploy.json]: updated

File: e-shop-website-with-ilb-ase/azuredeploy.json (hunk 11 above)
Comment (TylerLu) [reply e-shop-website-with-ilb-ase/azuredeploy.json]: all updated

File: e-shop-website-with-ilb-ase/metadata.json (hunk 16 above)
Comment (TylerLu) [reply e-shop-website-with-ilb-ase/metadata.json]: updated

File: e-shop-website-with-ilb-ase/prereqs/prereq.azuredeploy.json (hunk 18 above)
Comment (TylerLu) [reply e-shop-website-with-ilb-ase/prereqs/prereq.azuredeploy.json]: all updated

File: e-shop-website-with-ilb-ase/azuredeploy.json [hunk 26]
```diff
@@ -0,0 +1,888 @@
...
+    },
+    "sqlServerAdministratorLoginPassword": {
+      "type": "securestring",
+      "metadata": {
+        "description": "Administrator login password of the Azure SQL Server."
+      }
+    },
+    "azureAdTenantId": {
+      "type": "string",
+      "defaultValue": "",
```
[the remark this answers is not in the document]
Comment (TylerLu) [reply e-shop-website-with-ilb-ase/azuredeploy.json]: yes, it is required. Null is not a valid value. I removed the defaultValue.

File: e-shop-website-with-ilb-ase/azuredeploy.json (hunk 6 above)
Comment (TylerLu) [reply e-shop-website-with-ilb-ase/azuredeploy.json]: done

File: e-shop-website-with-ilb-ase/azuredeploy.json (hunk 13 above)
Comment (TylerLu) [reply e-shop-website-with-ilb-ase/azuredeploy.json]: done

File: e-shop-website-with-ilb-ase/azuredeploy.json (hunk 14 above)
Comment (TylerLu) [reply e-shop-website-with-ilb-ase/azuredeploy.json]: done

File: e-shop-website-with-ilb-ase/prereqs/prereq.azuredeploy.json (hunk 17 above)
Comment (TylerLu) [reply e-shop-website-with-ilb-ase/prereqs/prereq.azuredeploy.json]: removed

File: e-shop-website-with-ilb-ase/azuredeploy.parameters.json (hunk 15 above)
Comment (TylerLu) [reply e-shop-website-with-ilb-ase/azuredeploy.parameters.json]: done

**Comment by bmoore-msft** [comment]
Why are we skipping CI?

**Comment by TylerLu** [comment]
@bmoore-msft The deployment of ASE sometimes takes more than 2 hours and cuases timeout error. According to Christina's feedback below, we skipped CI.
> Templates and the automated tests are troublesome.  Our templates actually block the tests because of that.  There is a file named .ci_skip in there that does it.  You can look here: [https://github.com/Azure/azure-quickstart-templates/tree/master/201-web-app-asev2-create](https://github.com/Azure/azure-quickstart-templates/tree/master/201-web-app-asev2-create) for a reference.  That said, the very strong desire is that all our templates go through such testing.  The challenge with ASE is the long deployment times that are possible.  It should take < 1.5 hours.  Closer to 1.  Still, things do get stuck.  I personally haven’t got it to work with a template update yet but it hasn’t been at the top of my to do list.

**Comment by TylerLu** [comment]
@bmoore-msft How do you suggest about the hard code address prefixes of subnets and ip address of the Redis cache?
- 192.168.251.0/26
- 192.168.251.64/26
- 192.168.251.62

**Comment by bmoore-msft** [comment]
If there is one set of values that works for all users, then just put them as variables in the template - but 192 is a shared space, not public, so not sure how that will work...

**Comment by TylerLu** [comment]
@bmoore-msft Yes, the set of values works for all users if they follow the README.md. I have put them as variables.
These IP addresses are used **internally** in ASE, so 192 works.

**Pull request merged**